### ISSUE TITLE
setuid should be called in the child after forking the process in spawnProcess

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6443]
|[<img alt="GaretJax's avatar" src="https://avatars.githubusercontent.com/u/86236?s=50" width="50" height="50">](https://github.com/GaretJax)| @GaretJax reported|
|-|-|
|Trac ID|trac#6443|
|Type|defect|
|Created|2013-04-17 20:33:24Z|
|Branch|https://github.com/twisted/twisted/tree/only-child-setuid-6443|

Currently the Process class sets the euid in the parent process before forking (http://twistedmatrix.com/trac/browser/trunk/twisted/internet/process.py#L380)

There seem to be no reason this can't be more properly done in the forked child.

Attachments:

* [procmontest.tac](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7c8/7c83bdde3211f818c5e3e4ec9c7e46c5ed8304a6/5b031c6175570494fce4bce5031d59e25c2539ea.tac) (367 bytes) - added by exarkun on 2013-04-21 12:54:33Z - sudo PYTHONPATH=... twistd --euid --uid nobody --gid nogroup -ny procmontest.tac

<details><summary>Searchable metadata</summary>

```
trac-id__6443 6443
type__defect defect
reporter__GaretJax GaretJax
priority__normal normal
milestone__ 
branch__branches_only_child_setuid_6443 branches/only-child-setuid-6443
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1366230804000000 1366230804000000
changetime__1370031031000000 1370031031000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [38200]) Branching to 'only-child-setuid-6443'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=only-child-setuid-6443&num_builds=1).  Some slaves are broken right now, unfortunately.

Also worth noting that the test suite doesn't actually try to run a process with another uid/gid, since that requires root which our slaves don't generally have.  There are some tests with mocks, as apparent from the changes in the branch.  I did some manual testing with the attached tac.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Build results link above is wrong, sorry.  Here's a [working one](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/only-child-setuid-6443&num_builds=1).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince commented|
|-|-|

1.The docstrings for `test_mockSetUidInParent` and `test_mockPTYSetUidInParent` are no longer accurate.
   * It appears that 
 2. If we know how to test this, but it requires root, why don't we include the test but skipped.
   * `twisted.python.test.test_fakepwd.SPwdModuleTests` is an existing example of doing this.

Please merge after updating the docstring and adding a real test.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [38541]) Fix the docstrings of two tests which had their expectations modified.

refs [#6443](https://github.com/twisted/twisted/issues/6443)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

> * It appears that

Was this the start of an important comment, or an item you meant to delete?

> If we know how to test this, but it requires root, why don't we include the test but skipped.

Hm.  Okay, if we set up a slave that runs as root.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [38549]) pyflakes fixes

refs [#6443](https://github.com/twisted/twisted/issues/6443)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [38550]) Merge only-child-setuid-6443

Author: exarkun
Reviewer: tom.prince
Fixes: [#6443](https://github.com/twisted/twisted/issues/6443)

Avoid changing the process UID or GID when launching a child process which is to
run with a different UID or GID.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `reopened`|
|-|-|

(In [38552]) Revert r38550 - test suite regression

New tests fail on Windows, which lacks UIDs and GIDs.

Reopens: [#6443](https://github.com/twisted/twisted/issues/6443)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [38568]) Merge only-child-setuid-6443

Author: exarkun
Reviewer: tom.prince
Fixes: [#6443](https://github.com/twisted/twisted/issues/6443)

Avoid changing the process UID or GID when launching a child process which is to
run with a different UID or GID.
